File: caesar-cipher.erl
-module(prog).
-export([main/2]).
 
%% shift one char
cipherChar(Char,Key) when (Char >= $A) and (Char =< $Z) or
                   (Char >= $a) and (Char =< $z) ->
  Offset = $A + Char band 32,
  N = Char - Offset,
  Offset + (N + Key) rem 26;
 
%% handle large shift and negative.
key(Key) when Key < 0 ->
  26 + Key rem 26;
key(Key) when Key > 25 ->
  Key rem 26;
key(Key) ->
  Key.
 
main(inputString, Key) ->
  Encode = key(Key),
  Decode = key(-Key),
 
  io:format("inputString ----> ~s~n", [inputString]),
 
  ciphered = lists:map(fun(Char) -> cipherChar(Char, Encode) end, inputString),
  io:format("ciphered ---> ~s~n", [ciphered]),
 
  inputString = lists:map(fun(Char) -> cipherChar(Char, Decode) end, ciphered).

%% prog:main("helloWorld", 4).

File: caesarCipher.fs
open System

let shiftNum = 4
let inputString = "HelloWorld"

// perform cipher on one char, convert back to char
let ciph key num head =
  (((key + num - head) % 26) + head) |> char 

// read case and return 65 or 97
let case char = 
  if Char.IsLower char then 97 else 65
  
let cipher strIn numShift =
  let shift c =
    let asciiHead = case c
    ciph (int c) numShift asciiHead
  String.map shift strIn
 
// combine the cipher and encrypt methods
// remeber when you call cipher you must also provide a string!
let encrypt s k = cipher s k
let decrypt s k = cipher s (26 - k) 

let rec solve s acc =
    if acc = 26 then ""
    else 
      printfn "%s " s
      solve (encrypt s 1) (acc + 1)

let ciphered = encrypt inputString shiftNum 
let deciphered = decrypt ciphered shiftNum 

Console.WriteLine(inputString)
Console.WriteLine(ciphered)
Console.WriteLine(deciphered)

Console.WriteLine()

let solveObj = solve ciphered 0